src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+//!
+//! Implementation of [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki)
+//! signature message
+//!
+
+use consensus::Encodable;
+use hashes::{sha256, Hash};
+use std::{error, fmt, io};
+use util::taproot::{TapLeafHash, TapSighashHash};
+use {Script, SigHashType, Transaction, TxOut};
+
+/// Contains outputs of previous transactions to provide to the [SigHashCache::signature_hash]
+/// method, In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::Anyone] may be provided
+pub enum Prevouts {
+    /// When modifier `ANYONECANPAY` is provided, only prevout of the current input is needed
+    Anyone(usize, TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so he can reuse
+    /// the same data structure for multiple inputs
+    All(Vec<TxOut>),
+}
+
+const LEAF_VERSION_TAPSCRIPT: u8 = 0xc0;
+
+/// Information related to the script path spending
+pub struct ScriptPath {
+    script: Script,
+    code_separator: i32,
+    leaf_version: u8,
+}
+
+impl Default for ScriptPath {
+    fn default() -> Self {
+        ScriptPath {
+            script: Script::default(),
+            code_separator: -1,
+            leaf_version: LEAF_VERSION_TAPSCRIPT,
+        }
+    }
+}
+
+/// Possible errors in computing the signature message
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Should never happen since we are always encoding, thus we are avoiding wrap the IO error
+    IoError,
+
+    /// There are mismatches in the number of prevouts provided compared with the number of
+    /// inputs in the transaction
+    PrevoutsSize,
+
+    /// Requested a prevout index which is greater than the number of prevouts provided or a
+    /// [Prevouts::Anyone] with different index
+    PrevoutIndex,
+
+    /// A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`
+    PrevoutKind,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::IoError => write!(f, "IoError"),
+            Error::PrevoutsSize => write!(f, "Number of supplied prevouts differs from the number of inputs in transaction"),
+            Error::PrevoutIndex => write!(f, "The index requested is greater than available prevouts or different from the provided [Provided::Anyone] index"),
+            Error::PrevoutKind => write!(f, "A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`")
+        }
+    }
+}
+
+impl error::Error for Error {}
+
+impl Prevouts {
+    fn get_all(&self) -> Result<&[TxOut], Error> {
+        match self {
+            Prevouts::All(ref prevouts) => Ok(prevouts),
+            _ => Err(Error::PrevoutKind),
+        }
+    }
+
+    fn get(&self, input_index: usize) -> Result<&TxOut, Error> {
+        match self {
+            Prevouts::Anyone(index, prevout) => {
+                if input_index == *index {
```

Comment:
BIP-118 doesn;'t commit to the input_index, so this might not be robust?

Maybe make get take an Option index and only check if it's provided